src/rebar_paths.erl

```diff
@@ -217,22 +218,29 @@ get_apps(runtime, State) ->
     %% We get all project apps and for each of them we find
     %% their runtime deps (i.e., `applications' and `included_applications').
     ProjectApps = rebar_state:project_apps(State),
-    RuntimeApps =
-        [begin
-            Apps = rebar_app_info:applications(App),
-            IncludedApps = rebar_app_info:included_applications(App),
-            lists:foldl(
-              fun(AppName0, Acc) ->
-                  %% If the App is an atom convert it to a binary, otherwise leave it as it is.
-                  AppName1 = if is_atom(AppName0) -> atom_to_binary(AppName0, utf8);
-                                true -> AppName0
-                             end,
-                  %% We only care about those apps we could find in the `State'.
-                  case rebar_app_utils:find(AppName1, ProjectApps) of
-                      {ok, AppInfo} -> [AppInfo|Acc];
-                      error -> Acc
-                  end
-              end, [App], Apps ++ IncludedApps)
-         end || App <- ProjectApps],
-    %% We get rid of duplicated apps from the flattened list.
-    lists:usort(lists:flatten(RuntimeApps)).
\ No newline at end of file
+    AppsList = rebar_state:project_apps(State) ++ rebar_state:all_deps(State),
+    get_runtime_apps(ProjectApps, sets:new(), AppsList).
+
+get_runtime_apps([], RuntimeApps, _AppsList) -> sets:to_list(RuntimeApps);
+%% Skip kernel and stdlib, which are represented as atoms.
+get_runtime_apps([App|Rest], AppsAcc0, AppsList) when is_atom(App) ->
+    get_runtime_apps(Rest, AppsAcc0, AppsList);
+get_runtime_apps([App|Rest], AppsAcc0, AppsList) ->
+    Apps = rebar_app_info:applications(App),
+    IncludedApps = rebar_app_info:included_applications(App),
+    TotalApps = Apps ++ IncludedApps,
```

Comment:
Should probably drop the apps that are already in `AppsAcc0` so that we avoid re-processing the same transitive dependencies multiple times. Currently if A and B both rely on C and D, and that C also relies on D, I believe we can end processing C twice and D three times as well

Reply:
Gotcha